Add land use tutorial (#25677)
dotnet/docs: docs/machine-learning/tutorials/image-classification-model-builder.md

ms.topic: tutorial
ms.custom: mvc,mlnet-tooling
#Customer intent: As a non-developer, I want to use Model Builder to automatically train a model in Azure to classify images using Model Builder.
---

# Train an image classification model in Azure using Model Builder

Learn how to train an image classification model in Azure using Model Builder to categorize land use from satellite images.

This tutorial shows you how to create a C# class library to categorize land use based on satellite images with Model Builder.

In this tutorial, you:

> [!div class="checklist"]
>
> - Prepare and understand the data
> - Create a Model Builder config file
> - Choose a scenario
> - Load the data
> - Create an experiment in Azure
> - Train the model
> - Evaluate the model
> - Consume the model

> [!NOTE]
> Model Builder is currently in Preview.

## Prerequisites

- For a list of pre-requisites and installation instructions, visit the [Model Builder installation guide](../how-to-guides/install-model-builder.md).
- Azure account. If you don't have one, [create a free Azure account](https://aka.ms/AMLFree).
- ASP.NET and web development workload.

## Model Builder image classification overview

> This sample creates C# class library that categorizes land use from map satellite imagery using a deep learning model trained on Azure with Model Builder. The model itself is hosted as a web service in an ASP.NET Core Web API. You can find the source code for this tutorial in the [dotnet/machinelearning-samples](https://github.com/dotnet/machinelearning-samples/tree/master/samples/modelbuilder/ImageClassification_Azure_LandUse) GitHub repository

Reviewer: ```suggestion
This sample creates C# class library that categorizes land use from map satellite imagery using a deep learning model trained on Azure with Model Builder. You can find the source code for this tutorial in the [dotnet/machinelearning-samples](https://github.com/dotnet/machinelearning-samples/tree/master/samples/modelbuilder/ImageClassification_Azure_LandUse) GitHub repository
```

## Create a C# Class Library

> 1. Create a **C# Class Library** called "LandUse". Make sure **Place solution and project in the same directory** is **unchecked** (VS 2019).

Reviewer: ```suggestion
Create a **C# Class Library** called "LandUse". 
```

The EuroSAT dataset contains a collection of satellite images divided into ten categories (rural, industrial, river, etc.). The original dataset contains 27,000 images. For convenience, this tutorial only uses 2,000 of those images.

  ![Dataset](./media/image-classification-model-builder/dataset-images.png)

1. Download the subset of the [EuroSAT dataset](https://github.com/dotnet/machinelearning-samples/blob/master/samples/modelbuilder/ImageClassification_Azure_LandUse/assets.zip) and save it anywhere on your computer.
1. Unzip it.

## Create a Model Builder config file

When first adding Model Builder to the solution it will prompt you to create an `mbconfig` file. The `mbconfig` file keeps track of everything you do in Model Builder to allow you to reopen the session.

1. In Solution Explorer, right-click the **LandUse** project, and select **Add > Machine Learning Model...**.
1. In the dialog, name the Model Builder project **LandUse**, and click **Add**.

## Choose a scenario

![Model Builder Scenarios](../media/model-builder-scenarios.png)

To train your model, you need to select from the list of available machine learning scenarios provided by Model Builder.

> 1. For this sample, the task is image classification. In the scenario step of the Model Builder tool, select the **Image Classification** scenario.

Reviewer: ```suggestion
For this sample, the task is image classification. In the scenario step of the Model Builder tool, select the **Image Classification** scenario.
```

---
title: 'Tutorial: Train an image classification model in Azure using Model Builder'
description: Learn how to train an image classification model to recognize land use from satellite images in Azure using Model Builder
author: luisquintanilla
ms.author: luquinta
ms.date: 09/21/2021
ms.topic: tutorial
ms.custom: mvc,mlnet-tooling
#Customer intent: As a non-developer, I want to use Model Builder to automatically train a model in Azure to classify images using Model Builder.
---

# Train an image classification model in Azure using Model Builder

Learn how to train an image classification model in Azure using Model Builder to categorize land use from satellite images.

> This tutorial shows you how to create a Universal Windows Platform (UWP) application that hosts a model in an C# ASP.NET Core Web API to categorize land use based on satellite images.

Reviewer: It's no longer a UWP application. Update the description here.

In this tutorial, you:

> [!div class="checklist"]
>
> - Prepare and understand the data
> - Create a Model Builder config file
> - Choose a scenario
> - Load the data
> - Create an experiment in Azure
> - Train the model
> - Evaluate the model
> - Consume the model

> [!NOTE]
> Model Builder is currently in Preview.

## Prerequisites

> - For a list of pre-requisites and installation instructions, visit the [Model Builder installation guide](https://docs.microsoft.com/dotnet/machine-learning/how-to-guides/install-model-builder).

Reviewer: Use a relative path.

```suggestion
- For a list of pre-requisites and installation instructions, visit the [Model Builder installation guide](../how-to-guides/install-model-builder.md).
```

- Azure account. If you don't have one, [create a free Azure account](https://aka.ms/AMLFree).
- ASP.NET and web development workload.

## Model Builder image classification overview

> This sample creates a UWP application that categorizes land use from map satellite imagery using a deep learning model trained on Azure with Model Builder. The model itself is hosted as a web service in an ASP.NET Core Web API. You can find the source code for this tutorial in the [dotnet/machinelearning-samples](https://github.com/dotnet/machinelearning-samples/tree/master/samples/modelbuilder/ImageClassification_Azure_LandUse) GitHub repository

Reviewer: Since it's no longer a UWP app, update this description.

> ## Create solution

Reviewer: Since we won't be creating multiple apps, we can remove the solution section.

1. In Visual Studio, select **File > New > Project** from the menu bar.
1. In the New Project dialog, type "Solution" into the search box.
1. Select the **Blank Solution** template, then select the **Next** button.
1. In the **Name** text box, type "LandUse".
1. Select **Create**.

## Create a Console Project

> 

Reviewer: For this section, let's update the project type to a C# Class Library. Class Libraries make models more portable and you're not creating yet another console app when it's already one of the provided templates.

> 1. In Solution Explorer, right-click the **LandUse** solution, then select **Add > New Project**.

Reviewer: Since we're no longer working with solutions, you can remove the adding of a new project to the solution and instead you can create the project directly.

The EuroSAT dataset contains a collection of satellite images divided into ten categories (rural, industrial, river, etc.). The original dataset contains 27,000 images. For convenience, this tutorial only uses 2,000 of those images.

  ![Dataset](./media/image-classification-model-builder/dataset-images.png)

1. Download the subset of the [EuroSAT dataset](https://github.com/dotnet/machinelearning-samples/blob/master/samples/modelbuilder/ImageClassification_Azure_LandUse/assets.zip) and save it anywhere on your computer.
1. Unzip it.

## Create a Model Builder config file

When first adding Model Builder to the solution it will prompt you to create an `mbconfig` file. The `mbconfig` file keeps track of everything you do in Model Builder to allow you to reopen the session.

1. In Solution Explorer, right-click the **LandUse** project, and select **Add > Machine Learning Model...**.
1. In the dialog, name the Model Builder project **LandUse**, and click **Add**.

## Choose a scenario

![Model Builder Scenarios](../media/model-builder-scenarios.png)

To train your model, you need to select from the list of available machine learning scenarios provided by Model Builder.

> 1. For this sample, the scenario is image classification. In the scenario step of the Model Builder tool, select the **Image Classification** scenario.

Reviewer: ```suggestion
1. For this sample, the task is image classification. In the scenario step of the Model Builder tool, select the **Image Classification** scenario.
```

## Select an enviornment

Model Builder can run the training on different environments depending on the scenario that was selected.

> 1. Select "Azure" as your environment and click the **Set up workspace** button.

Reviewer: Since it's only one step, I think we can remove the number in this case.

```suggestion
Select **Azure** as your environment and click the **Set up workspace** button.
```

### Create experiment in Azure

An Azure Machine Learning experiment is a resource that needs to be created before running Model Builder training on Azure.

The experiment encapsulates the configuration and results for one or more machine learning training runs. Experiments belong to a specific workspace. The first time an experiment is created, its name is registered in the workspace. Any subsequent runs - if the same experiment name is used - are logged as part of the same experiment. Otherwise, a new experiment is created.

> 1. In the Create New Experiment dialog, choose your subscription from the **Subscription** dropdown.

Reviewer: Since it's only 1 step, we can remove the number.

```suggestion
In the Create New Experiment dialog, choose your subscription from the **Subscription** dropdown.
```

### Create workspace

> A workspace is an Azure Machine Learning resource that provides a central place for all Azure Machine Learning resources and artifacts created as part of training run.

Reviewer: ```suggestion
A workspace is an Azure Machine Learning resource that provides a central place for all Azure Machine Learning resources and artifacts created as part of a training run.
```

1. In the Create New Experiment dialog, select the **New** link next to the **Machine Learning Workspace name** dropdown.
1. In the Create A New Workspace dialog, type "landuse-wkspc" in the **Machine Learning Workspace name** text box.
1. Choose **East US** from the **Regions** dropdown. A region is the geographic location of the data center where your workspace and resources are deployed to. It is recommended that you choose a location close to where you or your customers are.
1. Select the **New** link next to the **Resource Groups** dropdown.
    1. In the Create New Resource Group dialog, type "landuse-rg" in the **Resource Group name** text box.
    1. Select **OK**.
1. Choose your newly created resource group from the **Resource Groups** dropdown.
1. Select **Create**.

    The provisioning process takes a few minutes. A request is made to Azure to provision the following cloud resources:

>     - Enterprise Azure Machine Learning workspace

Reviewer: I don't believe there's a distinction anymore so we can remove this.

```suggestion
    - Azure Machine Learning workspace
```

### Start training

Once you've configured your workspace and compute type, it's time to finish creating the experiment and start training.

1. Click the **Start Training** button.

    The training process takes some time and the amount of time may vary depending on the size of compute selected as well as amount of data. The first time a model is trained, you can expect a slightly longer training time because resources have to be provisioned. You can track the progress of your runs by selecting the **Monitor current run in Azure portal** link in Visual Studio.

    Throughout the training process, progress data is displayed in the Progress section of the train step.

    - Status displays the completion status of the training process.
    - Best accuracy displays the accuracy of the best performing model found by Model Builder so far. Higher accuracy means the model predicted more correctly on test data.
    - Algorithm displays the name of the best performing algorithm performed found by Model Builder so far.

1. Once training is complete, select the **Next step** button to move to evaluate the model.

### Evaluate the model

> The result of the training step is one model that had the best performance. In the evaluate step of the Model Builder tool, the **Details** tab in the output section, will contain the algorithm used by the best performing model in the **Algorithm** entry along with metrics in the **Best model Accuracy** entry.

Reviewer: Is the capitalization of "Best model Accuracy" correct?

Author: It's shown like this
![image](https://user-images.githubusercontent.com/1578160/134892993-a128182b-4db3-4a46-ab0d-7c7be2610c34.png)

I'm guessing we can just say "Best model accuracy" though or do you think there is a better way to word it?

Reviewer: Accuracy in the Best Model details

- For a list of pre-requisites and installation instructions, visit the [Model Builder installation guide](https://docs.microsoft.com/dotnet/machine-learning/how-to-guides/install-model-builder).
- Azure account. If you don't have one, [create a free Azure account](https://aka.ms/AMLFree).
- ASP.NET and web development workload.
- Universal Windows Platform development workload.

## Model Builder image classification overview

This sample creates a UWP application that categorizes land use from map satellite imagery using a deep learning model trained on Azure with Model Builder. The model itself is hosted as a web service in an ASP.NET Core Web API. You can find the source code for this tutorial in the [dotnet/machinelearning-samples](https://github.com/dotnet/machinelearning-samples/tree/master/samples/modelbuilder/ImageClassification_Azure_LandUse) GitHub repository

## Create solution

1. In Visual Studio, select **File > New > Project** from the menu bar.
1. In the New Project dialog, type "Solution" into the search box.
1. Select the **Blank Solution** template, then select the **Next** button.
1. In the **Name** text box, type "LandUse".
1. Select **Create**.

> ## Create a Console Project

Author: Simplified the tutorial by removing the UWP mentions and just showing the consumption from the Model Builder templates.
